### ISSUE TITLE
Clustering with TLS - Issues connecting to static route

### ISSUE THREAD

**Issue by ryancurtin** [nats-io/gnatsd#594]
I'm having a problem setting up a cluster with TLS, and while I'm fairly sure it's not a bug in gnatsd, it could be an opportunity to improve documentation and clarify usage (which I'm happy to do, once I can confirm if my setup is correct).  Anyway, I have a cluster with a seed instance and multiple client instances that will receive connections (behind a load balancer and within an auto-scaling group).  My configuration for the "seed" instance is as follows:
```
  listen: 0.0.0.0:4222
  https_port: 9223
  debug: true

  # Won't receive client connections, so this probably isn't necessary
  tls {
    cert_file: './wildcard_cert.pem'
    key_file: './wildcard_cert.key'
    timeout: 2
  }

  authorization {
    user: user
    password: password
  }

  cluster {
    listen: 0.0.0.0:5222
    tls {
      cert_file: './wildcard_cert.pem'
      key_file: './wildcard_cert.key'
      timeout: 2
    }
  }
```

"Client" instances are configured similarly, except they'll have a route to that particular seed instance and they use a self-signed certificate for the cluster route:
```
  listen: 0.0.0.0:4222
  https_port: 9223
  debug: true

  tls {
    cert_file: './wildcard_cert.pem'
    key_file: './wildcard_cert.key'
    timeout: 2
  }

  authorization {
    user:     ...
    password: ...
  }
  cluster {
    listen: 0.0.0.0:5222

    tls {
      cert_file: './server.crt'
      key_file: './server.key'
      ca_file: './ca.crt'
      timeout: 2
    }

    routes = [
      nats-route://seed.mydomain.com:5222
    ]
  }
```
Anyway, I wanted to confirm that my usage here was indeed correct (having a different certificate for "cluster" vs at the top level).  The actual issue I have is that the client handshake fails with the following message:
```
TLS route handshake error: x509: certificate signed by unknown authority
```

I've generated a self-signed certificate locally, similar to what you did in the tests.  The entire cluster works correctly there (as it always does).  In my production environment, the certificates are signed for the IP address of each container (with the ca.crt file generated beforehand to be shared across containers).  The wildcard cert referenced above is indeed valid, so it's much more likely it's an issue with my self-signed certificate.  

As this relates to documentation, I found this article to be helpful, but it's not actually reachable via the menu on the nats.io site:  [http://nats.io/documentation/server/gnatsd-tls/](http://nats.io/documentation/server/gnatsd-tls/).  I'd be glad to help add to the site with an example once I can confirm I'm on the right track here.  Please let me know if this makes sense, and if I can clarify anything further.  

#### Versions of `gnatsd` and affected client libraries used: 
1.0.2

#### OS/Container environment: 
Docker version 17.03.2-ce, build 7392c3b/17.03.2-ce (running in Amazon ECS)

**Comment by kozlovic** [comment]
@ryancurtin If you use self-signed certificates for the 'clients' instances, the CA needs to be known by the seed, not by the `clients`. You would need the ca certificate in the 'clients' if you used a self-signed certificate for the seed server. 

If each 'client' uses a different root ca, then the seed needs to have them all. Since `ca_file` is only one file, you would need to concatenate all root CAs of the `client` instances (their ca.crt files) into one.

Let me know if this helps.

**Comment by ryancurtin** [comment]
@kozlovic:  Thanks for your response!  From there, I was able to use dnsmasq to replicate a (mostly) realistic configuration and debug a bit faster.  Once I figure out the issue with certs it still wasn't working, so I dug a little deeper into what was different on production.  It turns out my actual problem, once I cleared up the cert issues, was the fact that I had the "seed" instance behind an ELB.  It was always getting the IP of the load balancer and never able to communicate with the "client" instances to form routes between them.  

I see that use case is covered by #447 so I will close this issue, as it looks like that's still in the process of making it onto the roadmap.  For anyone following the thread, my fix was to write Lambda function to make a DNS entry whenever the "seed" instance came online.  It's not ideal, but it seems to be working so far and it required minimal configuration changes.
